### PR TITLE
feat: Initial implementation of making startup tunes user configurable

### PR DESCRIPTION
**Pull request by saidinesh5**
Right now we represent a startup tune using:

`Eep_Pgm_Startup_Tune` - An array of 128 bytes to store the startup melody

Bluejay Startup Melody has a structure like:

| [bpm]   | [default octave] | [default duration]  | [62 (Temp4, Temp3) values] |
| ------- | :--------------: | :-----------------: | :------------------------: |
| 2 bytes |      1 byte      |       1 byte        |   124 bytes                |

In general,
* The first 4 bytes of the byte array are considered metadata and are ignored by the parer in the ESC firmware.
* Temp3 is the pulse duration - a measure of the frequency of the note. Temp 3 == 0 => This is a pause note.
* Temp4 is the number of pulses - a measure of the duration of the note. Temp4 == 0 => Terminate the startup melody loop

We then blindly loop through these pulses and call the wait/beep subroutine.

This should give us 62 musical notes, which I think should be enough for
a start up routine.

The big advantage of doing it this way is that the bluejay configurator
app does most of the logic of converting music to these tuples, and the
whole implementation will be possible in under 200 bytes of firmware space

The main question now is: are 62 notes enough?

Fixes https://github.com/mathiasvr/bluejay/issues/3

Requires: https://github.com/mathiasvr/bluejay-configurator/pull/13

### CODE REVIEW

File: Bluejay.asm [hunk 1]
```diff
@@ -3539,18 +3541,38 @@ pgm_start:
...
+	mov	Temp3,	A				; TODO: See if flash eeprom can be read directly into Temp3
+	jz	startup_melody_item_wait_ms	; TODO: Have to see why these having zero notes is bricking the ESC
+	inc	DPTR
+	call	beep
+	sjmp	startup_melody_loop_next_note
+
+startup_melody_item_wait_ms:
+	mov	A,	Temp4
+	mov	Temp2,	A
+	call	wait_ms_o
```
Comment (mathiasvr) [inline Bluejay.asm:3571]: The call to `wait_ms_o` mutates `Temp1` so it should not be used as counter.
Notice how the general pattern is that registers are not pushed to the stack, so be careful with jumps and calls.

Comment (saidinesh5) [reply Bluejay.asm:3570]: Good catch. Pushed the fixes. The ESC is now able to play the current start up tune properly. Still couldn't rescue my bricked one. Probably my bad soldering skills are to blame, will look at it later. For now i have 2 more ESCs to brick.

The next todo I have already implemented but haven't pushed is:

* Remove the additional setting to select between startup beep vs. startup tune as we can use the startup tune for everything. Should i push this change? That will cause a change in Eeprom layout. Or do you want to reuse that variable space for something else?

The main todo i am still figuring out how to is:

Make the `Eep_Pgm_Startup_Tune` configurable by the configurator.

* Why are eeprom read/write functions split into block 1/block 2? Is there any limit to the amount of data that can be read/written at once?
* I keep getting `*** WARNING L30: MEMORY SPACE OVERLAP` when i moved the startup tune array before the `Eep_Dummy` (even after adjusting the next CSEG's location.) What am i missing?

Can you share any tips on how to implement this task?

Comment (mathiasvr) [reply Bluejay.asm:3570]: - I'm planning to change the eeprom layout but have left it mostly backward compatible with BLHeli for now since there are a number of assumptions about it in the code and the configurator. We probably need to decide if we want to merge this beep melody feature without too many eeprom changes, or if we should wait merging it until we have a new eeprom layout structure.
- The read/write blocks are probably legacy code from blheli. I think maybe block 2 was only used on some escs? Also some mcus with real eeprom might have had different limitations.
- Maybe related to reserved space in `Common.inc` at `1A40h` and `1A50h`? This is part of what I'm planning to change as well.

Comment (saidinesh5) [reply Bluejay.asm:3570]: >  We probably need to decide if we want to merge this beep melody feature without too many eeprom changes, or if we should wait merging it until we have a new eeprom layout structure.

I don't mind it either way, but some bugs may pop up only after you integrate the "make this configurable with configurator" feature, (like total byte count of parameters + tune becoming > 255 and then having to split things up). So reserving specific address for the melody data seems to make more sense. 

For the code part, it seems like making a change for a very specific use case and would be nice to have something more "generic/extensible" like "ESC variable parameters" or some such thing, which just reserves some writable buffer that can be filled with values. Then we can directly reserve like 250 bytes for everything someone may want to add in the future.

Comment (saidinesh5) [reply Bluejay.asm:3570]: @mathiasvr Any updates on the eeprom layout changes? Will that come on top of the layout related refactoring?

Comment (mathiasvr) [reply Bluejay.asm:3570]: The eeprom layout changes are actually a separate thing. The esc layout refactoring is a bit overkill but was something I was working on while trying to support a new type of esc pinout.

I'm not sure when to change eeprom layout because I would like to have settings mostly stable at that point (and do it only once). I hope you're not dependant on that? I'm fine if you reserve an address for the melody in this pr.

Comment (saidinesh5) [reply Bluejay.asm:3570]: Ah will try to implement read out of melody data via. configurator in this PR itself then. I think this will be a busy week for me, so most probably I will look into it during the weekend.

Comment (mathiasvr) [reply Bluejay.asm:3570]: Ah sorry, I forgot you wanted to look into the configurator already. I will look into the eeprom layout soon, I think many of the changes can be done and still be compatible. I can try to prepare the read/write of melody data as well?

Comment (saidinesh5) [reply Bluejay.asm:3570]: Hi, just a heads up. This weekend i looked at your forked blheli configurator code and started adding a Melody component to it. I think the interesting thing is i don't think we have to add any more changes to the ESC code to make it send the melody data specifically.

What's happening is:

`readbackSettings = (await _4way.read(BLHELI_SILABS_EEPROM_OFFSET, BLHELI_LAYOUT_SIZE)).params;`

So just updating BLHELI_LAYOUT_SIZE should work i think.

Yet to test it out, and my knowledge of react is also non existent at this point but this is turning out to be an interesting challenge so far. Learning a lot of things I wouldn't have usually touched.

Let me know if i need to be careful about more things. I think next week onwards my schedule should be more free to focus on these things better.

Comment (mathiasvr) [reply Bluejay.asm:3570]: I have been working a bit on this as well. Looks like we can only write 256 bytes at a time, so might be easiest to start with melodies within that range.

Comment (saidinesh5) [reply Bluejay.asm:3570]: But 256 bytes is plenty right?

image

I have some local changes i am yet to push (increasing the melody length from 100 bytes to 128 bytes - so we get 64 notes), early end the loop if the current note duration is 0. Will push them once i test this out.

Finally managed to get the configurator built - it was missing some OS X dependency (i am on linux). After moving it to optional dependency it built fine.

**Comment by mathiasvr** [comment]
This is not important right now but I will just share here that I think it would be nice if we could use an existing "melody syntax" in the configurator.

@stylesuxx introduced me to [RTTTL](https://en.wikipedia.org/wiki/Ring_Tone_Transfer_Language) and I see there is also [MML](https://en.wikipedia.org/wiki/Music_Macro_Language).

**Comment by saidinesh5** [comment]
@mathiasvr Yup. Already started extending https://github.com/adamonsoon/rtttl-parse to spit out Temp3, Temp4 values. Also I fixed the issues i was having with my React component. I was trying to base it on the Number component which uses:

```js
handleChange: function (component, value){}
``` and `value` was always undefined.

So just went with something from react documentation and everything started working.
```js
handleChange: function(e) {
        this.props.onChange(e.target.name, parseInt(e.target.value));
}
```

Without further ado, the obligatory screenshot:

image

Now, do you happen to know any quick formula for me to convert a sound frequency value to the Temp3 register value of beep subroutine? Will save me a lot of time with my rtttl-parse functions.

Basically I have 2 problems left to solve now:

1) Parse rtttl notes into `Eep_Pgm_Startup_Tune` array
2) Parse `Eep_Pgm_Startup_Tune` into rtttl notes for easy editing. The latter seems to be tricky because there is a many to one mapping of rtttl notes ->  `Eep_Pgm_Startup_Tune` array, because of having to split up values > 255.

**Comment by mathiasvr** [comment]
@saidinesh5 Nice work, looks good.

> Now, do you happen to know any quick formula for me to convert a sound frequency value to the Temp3 register value of beep subroutine? Will save me a lot of time with my rtttl-parse functions.

I can try to see if I can come up with something.

> Basically I have 2 problems left to solve now:
> 
> 1. Parse rtttl notes into `Eep_Pgm_Startup_Tune` array
> 2. Parse `Eep_Pgm_Startup_Tune` into rtttl notes for easy editing. The latter seems to be tricky because there is a many to one mapping of rtttl notes ->  `Eep_Pgm_Startup_Tune` array, because of having to split up values > 255.

What do you mean by splitting up values?

**Comment by mathiasvr** [comment]
Based on some old data I had, I made this very crude approximation:
```js
Temp3 = Math.round(328687 / 496**1.3746)
```

Maybe it can be used as a starting point.

**Comment by saidinesh5** [comment]
@mathiasvr wow. That's a surprisingly good approximation. Needs a little tuning but can do that later by brute forcing all possible Temp3 values and map them to the note frequencies. Shouldn't be too difficult, since we are dealing with an 8 bit word.
And splitting up values as in, this is the algorithm I am using to convert rtttl to the startup array:

````js
/**
 * Parse RTTTL to Bluejay ESC startup tone data
 *
 * @param {string} rtttl - RTTTL String
 * @returns an array of (Number of pulses, Pulse width) tuples
 */
static parseToBluejayStartupTone(rtttl, startupToneLength) {
  let parsedData = Rtttl.parse(rtttl);
  startupToneLength = startupToneLength || 128;
  
  const MAX_ITEM_VALUE = 2**8;
  // Melody is basically an array of [{ duration(in ms): number, frequency (in Hz): number }]
  let melody = parsedData.melody;  
  let result = new Uint8Array(startupToneLength);

  function frequencyToBluejayTemp3(freq) {
  // TODO: Later on make this more accurate using brute force
    return Math.round(328687/freq**1.3746);
  }

  var currentResultIndex = 0;
  var currentMelodyIndex = 0;

  while(currentMelodyIndex < melody.length && currentResultIndex < result.length) {
    var item = melody[currentMelodyIndex];
    if (item.frequency != 0) {
        // temp3 is a measure of wavelength of the sound
        let temp3 = frequencyToBluejayTemp3(item.frequency);

        if (temp3 > 0 && temp3 < MAX_ITEM_VALUE) {
            let duration_per_pulse_ms = 1000/item.frequency; //(25 + temp3*200*25/150)/1000;
            let pulses_needed = Math.round(item.duration / duration_per_pulse_ms);
            
            while (pulses_needed > 0 && currentResultIndex < result.length) {
                result[currentResultIndex++] = pulses_needed%MAX_ITEM_VALUE;
                result[currentResultIndex++] = temp3;
                pulses_needed = Math.floor(pulses_needed/MAX_ITEM_VALUE);
            }
        } else {
            console.warn("Skipping note of frequency: ", item.frequency)
        }
    } else {
        // Can wait from 1-255ms for each (Temp3, Temp4) tuple
        // So split up a single silent note, if we have to
        let duration = Math.round(item.duration);

        while (duration > 0 && currentResultIndex < result.length) {
            result[currentResultIndex++] = duration%MAX_ITEM_VALUE;
            result[currentResultIndex++] = 0;
            duration = Math.floor(duration/MAX_ITEM_VALUE);
        }
    }
    
    currentMelodyIndex++;
  }
  
  if (currentMelodyIndex < melody.length) {
      console.warn("Only " + currentMelodyIndex + " notes out of " + melody.length + " fit in the startup sequence");
  }
    
  return result
}
````

If the duration can't fit in one Temp4 variable, we split a single rtttl note into multiple (Temp4, Temp3) tuples. That means we lose the original Rtttl data during this conversion process. On top of that, if a given 1/4th note duration can be achieved by a tempo of x bpm, we can create the same note duration using a 1/8th note and double the tempo.

So right now my inverse function looks like:
````js
static parseBluejayStartupToneToRtttl(startupTone) {
    // First glob together items we had to split up due to the 8 bit limit
    
    // Convert the array of [(Temp4, Temp3)] to an array of [{duration: number, frequency: number}] . This is kind of like the coin change problem now. But have to first find out what the coins are, before we can go down the path of a greedy approach

    //return rttf format as a string: 'Name:Defaults:rtttl'
}
````

**Comment by mathiasvr** [comment]
@saidinesh5 It's possible to change the range of Temp3/Temp4 if preferable, I already did this once before. We could also encode the duration differently to avoid duplicating notes? Anyway, looks good so far, do you know what is the typical range of frequencies/notes and durations we need to support?

I can also make a more accurate formula by measuring the frequency produced by the beep routine. This should be easier than brute force even though i'm not entirely sure what you had in mind.

Also, I shipped a version of the configurator that uses eeprom revision `202` for something else, my bad.

**Comment by saidinesh5** [comment]
Oh that's interesting. Originally i was trying to write the decoder completely in assembly but it got annoying very quickly to look up frequencies and calculate note durations in assembly, so thought we can do it nicely in configurator, with proper error checking - that way the ESC can blindly play what configurator gives it.

These are the tempos (in beats per minute) rtttl tones usually use:

```[
      '25', '28', '31', '35', '40', '45', '50', '56', '63', '70', '80', '90', '100',
      '112', '125', '140', '160', '180', '200', '225', '250', '285', '320', '355',
      '400', '450', '500', '565', '635', '715', '800', '900'
]```

So the shortest note is: `1/32 * (60000/900)` ~ 2ms
And the longest note is: `3/2 * (60000/25)` = 3600ms

As for revision 202 - no worries. Can change mine to 203. Have to squash this branch and rename commits to make this branch mergable anyway.

I opened a pull request of the configurator and added some TODOs there in case you want to look into this: https://github.com/mathiasvr/bluejay-configurator/pull/13

File: Bluejay.asm [hunk 2]
```diff
@@ -3540,21 +3544,40 @@ ENDIF
...
+	mov	Temp5,	#40h
+	mov	DPTR,	#Eep_Pgm_Startup_Tune
+
+startup_melody_loop:
+	; Read current location at Eep_Pgm_Startup_Tune to Temp4 and increment DPTR. If the value is 0, no point trying to play this note
+	clr	A
+	movc	A,	@A+DPTR
+	inc	DPTR
+	mov	Temp4,	A
+	jz	startup_beep_done
```
Comment (mathiasvr) [inline Bluejay.asm:3560]: This is supposed to only skip one note right?
```suggestion
	jz	startup_melody_loop_next_item
```

Comment (mathiasvr) [reply Bluejay.asm:3560]: Ah is it because we only expect to see zero when the melody is over?

Comment (saidinesh5) [reply Bluejay.asm:3560]: Yep. More like not let it run through the rest of the loop if there is a zero duration note. Originally, the idea was to use the space from  unused notes for storing some metadata, like tempo etc.. in case we need them in the configurator

**Comment by mathiasvr** [comment]
I have looked more into rtttl and I think the range of allowed durations should be calculated like this:

Shortest: `60000 / 900 * 4/32` = 8.3ms
Longest: `60000 / 25 * 4` = 9.6s

I might try to update the beep routine to better accommodate musical notes to simplify how we process and store notes/timings.
